agent/ui_endpoint.go

```diff
@@ -389,40 +413,53 @@ RPC:
 	return topo, nil
 }
 
-func summarizeServices(dump structs.ServiceDump, cfg *config.RuntimeConfig, dc string) (map[structs.ServiceName]*ServiceSummary, map[structs.ServiceName]bool) {
+func summarizeServices(dump structs.ServiceDump, cfg *config.RuntimeConfig, dc string) (map[structs.PeeredServiceName]*ServiceSummary, map[structs.PeeredServiceName]bool) {
 	var (
-		summary  = make(map[structs.ServiceName]*ServiceSummary)
-		hasProxy = make(map[structs.ServiceName]bool)
+		summary  = make(map[structs.PeeredServiceName]*ServiceSummary)
+		hasProxy = make(map[structs.PeeredServiceName]bool)
 	)
 
-	getService := func(service structs.ServiceName) *ServiceSummary {
-		serv, ok := summary[service]
+	getService := func(psn structs.PeeredServiceName) *ServiceSummary {
+		serv, ok := summary[psn]
 		if !ok {
 			serv = &ServiceSummary{
-				Name:           service.Name,
-				EnterpriseMeta: service.EnterpriseMeta,
+				Name:           psn.ServiceName.Name,
+				EnterpriseMeta: psn.ServiceName.EnterpriseMeta,
 				// the other code will increment this unconditionally so we
 				// shouldn't initialize it to 1
 				InstanceCount: 0,
+				PeerName:      psn.Peer,
 			}
-			summary[service] = serv
+			summary[psn] = serv
 		}
 		return serv
 	}
 
 	for _, csn := range dump {
+		var peerName string
+		// all entities will have the same peer name so it is safe to use the node's peer name
+		if csn.Node != nil {
+			peerName = csn.Node.PeerName
+		} else {
+			peerName = structs.DefaultPeerKeyword
+		}
```

Comment:
This will never be nil, and if it was there's a bunch of other places below that would panic
```suggestion
		peerName = csn.Node.PeerName
```

Reply:
When building a CheckServiceNode we start from a service instance (`structs.ServiceNode`), and then fetch its node (`structs.Node`). An invariant in the catalog is that services can't exist without a node, so the node for a service that exists won't be nil.

Reply:
I can appreciate that; I was thinking more of the lines of defense in depth and such but I can take this out

Reply:
actually it seems that UIGateway tests call this summarize func with nil  `cfn.Node`; not sure if feature or bug

Reply:
https://app.circleci.com/pipelines/github/hashicorp/consul/34387/workflows/03191cf7-b29c-45d0-ad6d-a5d6845e46e6/jobs/747784

Reply:
For instance, looking at some test setups --> 

```
&{Node:<nil> Service:<nil> Checks:[] GatewayService:0xc000e24620} 
```

Reply:
Ah ok. Yep I'm seeing the `continue` in L457 now.